server/src/daemon/service.rs

```diff
@@ -0,0 +1,311 @@
+        Algorithm,
+    },
+    portal::{Item, Keyring},
+    Key,
+};
+use tokio::sync::RwLock;
+use zbus::{
+    proxy::ProxyDefault,
+    zvariant::{ObjectPath, OwnedObjectPath, OwnedValue, Value},
+    ObjectServer, SignalContext,
+};
+
+use super::{
+    collection::Collection, error::ServiceError, prompt::Prompt, service_manager::ServiceManager,
+    session::Session, Result,
+};
+
+#[derive(Debug)]
+pub struct Service {
+    collections: RwLock<Vec<Collection>>,
+    keyring: Arc<Keyring>,
+    cnx: Mutex<Option<zbus::Connection>>,
+    manager: Arc<Mutex<ServiceManager>>,
+    sessions_counter: RwLock<i32>,
+}
+
+#[zbus::interface(name = "org.freedesktop.Secret.Service")]
+impl Service {
+    pub async fn open_session(
+        &self,
+        algorithm: Algorithm,
+        input: Value<'_>,
+        #[zbus(object_server)] object_server: &zbus::ObjectServer,
+    ) -> Result<(OwnedValue, OwnedObjectPath)> {
+        let (service_public_key, aes_key) = match algorithm {
+            Algorithm::Plain => (None, None),
+            Algorithm::Encrypted => {
+                let private_key = Key::generate_private_key();
+                let client_public_key = Key::from(input);
+                (
+                    Some(Key::generate_public_key(&private_key)),
+                    Some(Key::generate_aes_key(&private_key, &client_public_key)),
+                )
+            }
+        };
+
+        *self.sessions_counter.write().await += 1;
+        let session = Session::new(
+            aes_key,
+            Arc::clone(&self.manager),
+            *self.sessions_counter.read().await,
+        );
+        self.manager
+            .lock()
+            .unwrap()
+            .insert_session(session.path().to_owned(), session.to_owned());
+
+        object_server
+            .at(session.path().to_owned(), session.to_owned())
+            .await?;
+
+        let key = service_public_key
+            .map(|k| OwnedValue::from(&k))
+            .unwrap_or_else(|| Value::new::<Vec<u8>>(vec![]).try_to_owned().unwrap());
+
+        Ok((key, session.path().into()))
+    }
+
+    #[zbus(out_args("collection", "prompt"))]
+    pub async fn create_collection(
+        &self,
+        #[zbus(signal_context)] ctxt: SignalContext<'_>,
+        properties: Properties,
+        alias: &str,
+        #[zbus(object_server)] object_server: &ObjectServer,
+    ) -> Result<(OwnedObjectPath, Prompt)> {
+        let collection = Collection::new(
+            properties.label(),
+            alias,
+            SystemTime::now()
+                .duration_since(SystemTime::UNIX_EPOCH)
+                .unwrap(),
+            Arc::clone(&self.keyring),
+            Arc::clone(&self.manager),
+        );
+        self.collections.write().await.push(Collection::new(
+            collection.label(),
+            alias,
+            *collection.created(),
+            Arc::clone(&self.keyring),
+            Arc::clone(&self.manager),
+        ));
```

Comment:
You already create the same collection above? this is wrong.

Reply:
So, the reason why I'm doing this is, I have to update the `self.collections` [Vec](https://github.com/bilelmoussaoui/oo7/blob/b2abf3c42aa967882e3e34172c82ff06e3bbb608/server/src/daemon/service.rs#L31) when a new collection is created right. Since we need this just created `collection` later at `object_server.at()`, we can't just passed to it `push()`. So, we have to clone it. But, again `Collection` can't implement the `Clone` trait (Clone is not implemented for tokio ...). So, I can't do, `self.collections.write().await.push(collection.clone())`.

To fix this I have tried to clone the just created collection by hand using the values retrieved from getters :)

Reply:
So, here is the full error,

```
error[E0277]: the trait bound `tokio::sync::RwLock<Vec<item::Item>>: Clone` is not satisfied
  --> server/src/daemon/collection.rs:27:5
   |
24 | #[derive(Clone, Debug)]
   |          ----- in this derive macro expansion
...
27 |     pub(crate) items: RwLock<Vec<super::item::Item>>,
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Clone` is not implemented for `tokio::sync::RwLock<Vec<item::Item>>`
   |
   = note: this error originates in the derive macro `Clone` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `tokio::sync::RwLock<std::string::String>: Clone` is not satisfied
  --> server/src/daemon/collection.rs:28:5
   |
24 | #[derive(Clone, Debug)]
   |          ----- in this derive macro expansion
...
28 |     alias: RwLock<String>,
   |     ^^^^^^^^^^^^^^^^^^^^^ the trait `Clone` is not implemented for `tokio::sync::RwLock<std::string::String>`
   |
   = note: this error originates in the derive macro `Clone` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `AtomicBool: Clone` is not satisfied
  --> server/src/daemon/collection.rs:30:5
   |
24 | #[derive(Clone, Debug)]
   |          ----- in this derive macro expansion
...
30 |     locked: AtomicBool,
   |     ^^^^^^^^^^^^^^^^^^ the trait `Clone` is not implemented for `AtomicBool`
```

Reply:
Wrap those in Arcs. Nothing new here.

Reply:
Ha! :) will do, thanks!

Reply:
Done.